sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+
+            await using var connection = CreateConnection();
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                Stopwatch cycleDuration = Stopwatch.StartNew();
+
+                // Renew this instance's ownership so they don't expire.
+                // TODO: renew only after retrieving updated ownership so we use updated checkpoints.
+
+                await RenewOwnershipAsync().ConfigureAwait(false);
+
+                // From the storage service provided by the user, obtain a complete list of ownership, including expired ones.  We may still need
+                // their eTags to claim orphan partitions.
+
+                var completeOwnershipList = (await ListOwnershipAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup)
+                    .ConfigureAwait(false))
+                    .ToList();
+
+                // Filter the complete ownership list to obtain only the ones that are still active.  The expiration time defaults to 30 seconds,
+                // but it may be overridden by a derived class.
+
+                IEnumerable<PartitionOwnership> activeOwnership = completeOwnershipList
+                    .Where(ownership => DateTimeOffset.UtcNow.Subtract(ownership.LastModifiedTime.Value) < OwnershipExpiration);
+
+                // Dispose of all previous partition ownership instances and get a whole new dictionary.
+
+                InstanceOwnership = activeOwnership
+                    .Where(ownership => ownership.OwnerIdentifier == Identifier)
+                    .ToDictionary(ownership => ownership.PartitionId);
+
+                // Some previously owned partitions might have had their ownership expired or might have been stolen, so we need to stop
+                // the processing tasks we don't need anymore.
+
+                await Task.WhenAll(ActivePartitionProcessors.Keys
+                    .Except(InstanceOwnership.Keys)
+                    .Select(partitionId => StopPartitionProcessingIfRunningAsync(partitionId, ProcessingStoppedReason.OwnershipLost)))
+                    .ConfigureAwait(false);
+
+                // Now that we are left with processing tasks that should be running, check their status.  If any has stopped, it
+                // means a failure has happened, so try closing it and starting a new one.  In case we don't have a task that should
+                // exist, create it.  This might happen if the user hasn't updated ActivePartitionProcessors when initializing processing
+                // in the previous cycle.
+
+                await Task.WhenAll(InstanceOwnership
+                    .Select(async kvp =>
+                    {
+                        if (!ActivePartitionProcessors.TryGetValue(kvp.Key, out Task processingTask) || processingTask.IsCompleted)
+                        {
+                            await StopPartitionProcessingIfRunningAsync(kvp.Key, ProcessingStoppedReason.Exception).ConfigureAwait(false);
```

Comment:
Should you remove these using `TryRemove` instead of getting them here?

Reply:
We not necessarily want to remove the tasks. We keep them if they have not stopped. Moreover, `StopPartitionProcessingIfRunningAsync` captures any exception thrown by the task, delegates it to the error handler (not yet, but it will), and calls the close handler.